docs/custom_promise_types/modules.md

```diff
@@ -0,0 +1,367 @@
+    * The change should be explained in a `info` level log message
+  * `not_kept` - promise not satisfied before, and could not be fixed
+    * Should be explained by an `error` level log message
+* Teminate:
+  * `success` - Module succesfully terminated without errors
+  * `failure` - There were problems when trying to clean up / terminate
+    * Should be explained by a `critical` level log message
+
+Built-in CFEngine promises may have multiple outcomes when evaluated.
+For the sake of simplicity, custom promises may only have 1 outcome, logic should be as follows:
+
+1. If the promise was already satisfied and no changes were necessary, the promise was `kept`
+2. Otherwise, if there were issues and the module could not fix/satisfy all parts of the promise, it was `not_kept`.
+3. Otherwise, if some changes were made (successfully), the promise was `repaired`.
+
+### Protocol
+
+To implement a promise module from scratch, you will have to implement the promise module protocol.
+2 variants of the protocol are provided.
+1 JSON based and 1 line based.
+
+**Note: if you are using the library, as shown above, you don't have to care / learn about the following internals of the protocol.**
+
+### Protocol header
+
+When a promise module starts, the agent sends a protocol header (single line), followed by 2 newline characters.
+The promise module responds with a protocol header (single line), followed by 2 newline characters.
+
+The protocol header contains:
+
+* name of program
+* version of program
+* version of protocol
+* space separated list of features available (optional)
+
+The header has the same syntax regardless of protocol (It is used to determine protocol).
+
+#### Example headers:
+
+Request sent by cf-agent to promise module:
+
+```
+CFEngine 3.16.0 v1
+```
+
+Response, requesting line based protocol:
+
+```
+git_promises 0.0.1 v1 line_based
+```
+
+Response, requesting JSON based protocol:
+
+```
+git_promises 0.0.1 v1 json_based
+```
+
+After the initial exchange of protocol headers, the module responds to requests from the agent, using the agreed upon protocol.
+Requests and responses follow the same format and rules, and are commonly called messages.
+For each request, there should be exactly one response.
+
+### Protocol request sequence
+
+A promise module will receive 3 types of requests (operations):
+
+1. `validate_promise` - Validate the names and types of promise attributes
+2. `evaluate_promise` - Perform changes to the system
+3. `terminate` - Shut down the module
+
+The module should not make any assumptions about the order of validation or evaluation.
+Each promise might be validated and evaluated, 0, 1, or more times, in any order.
+In general, the promise module should not require that state is maintained between the individual requests.
+
+The agent _might_ start multiple processes of the same module in parallel, for example for a module implementing multiple promise types.
+The module should not rely on this.
```

Comment:
> I suppose it can never spawn a module for the same promise though? I think this should be specified.

I don't see a reason why we would, but I also don't think it should be problematic. cf-agent is single threaded, the modules running in parallell should just be waiting for another request, they should finish the work on one request before responding, and there shouldn't be any state / locks persisting after the requests. I was not intending for us to send multiple requests to different modules / instances of a module and allowing them to actually perform changes in parallell.

Reply:
Ok, understood, that makes it much easier. But then I would mention the following: "A module must not hold any locks while waiting for a new request".

Reply:
Done.